From eec38a200357b195efbb23bb645ab721c040f246 Mon Sep 17 00:00:00 2001
From: Vicente Olivert Riera <Vincent.Riera@imgtec.com>
Date: Thu, 3 Nov 2016 12:59:39 +0000
Subject: [PATCH] iptunnel.c: do not include netinet/ip.h

This fixes a struct redefinition problem like this one:

================================
In file included from /usr/include/linux/if_tunnel.h:6:0,
                 from iptunnel.c:34:
/usr/include/linux/ip.h:85:8: error: redefinition of 'struct iphdr'
 struct iphdr {
        ^
In file included from iptunnel.c:29:0:
/usr/include/netinet/ip.h:45:8: note: originally defined here
 struct iphdr
        ^
================================

iptunnel.c includes netinet/ip.h which contains a definition of the
iphdr struct.

iptunnel.c also includes linux/if_tunnel.h which includes linux/ip.h
which contains a definition of the iphdr struct.

So, both netinet/ip.h and linux/ip.h define the iphdr struct, and both
of them have been included directly or indirectly by iptunnel.c. Because
of that the compilation fails due to a struct redefinition.

The problem can be solved by just not including netinet/ip.h.

However, some Linux headers included in certain toolchains may not have
an updated linux/if_tunnel.h which includes linux/ip.h, so we need to
include it unconditionally otherwise linux/if_tunnel.h will use the
struct iphdr before being defined and the compilation will also fail in
this way:

================================
In file included from iptunnel.c:33:0:
/usr/include/linux/if_tunnel.h:37:16: error: field 'iph' has incomplete type
  struct iphdr  iph;
                  ^
================================

Upstream status: merge request sent
https://sourceforge.net/p/net-tools/code/merge-requests/4/

Signed-off-by: Vicente Olivert Riera <Vincent.Riera@imgtec.com>
---
 iptunnel.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/iptunnel.c b/iptunnel.c
index 3fe1b84..e2ec2d8 100644
--- a/iptunnel.c
+++ b/iptunnel.c
@@ -26,11 +26,11 @@
 #include <sys/socket.h>
 #include <sys/ioctl.h>
 #include <netinet/in.h>
-#include <netinet/ip.h>
 #include <arpa/inet.h>
 #include <net/if.h>
 #include <net/if_arp.h>
 #include <linux/types.h>
+#include <linux/ip.h>
 #include <linux/if_tunnel.h>

 #include "config.h"
-- 
2.10.1

